sklearn/tests/test_metaestimators_metadata_routing.py

```diff
@@ -17,109 +22,279 @@
-    ids=[str(x) for x in get_empty_metaestimators()],
+    METAESTIMATORS,
+    ids=[str(row["metaestimator"].__class__.__name__) for row in METAESTIMATORS],
 )
 def test_default_request(metaestimator):
     # Check that by default request is empty and the right type
-    assert_request_is_empty(metaestimator.get_metadata_routing())
-    assert isinstance(metaestimator.get_metadata_routing(), MetadataRouter)
+    cls = metaestimator["metaestimator"]
+    estimator = metaestimator["estimator"]()
+    estimator_name = metaestimator["estimator_name"]
+    instance = cls(**{estimator_name: estimator})
+    assert_request_is_empty(instance.get_metadata_routing())
+    assert isinstance(instance.get_metadata_routing(), MetadataRouter)
 
 
 @pytest.mark.parametrize(
-    "MultiOutput, Estimator",
-    [
-        (MultiOutputClassifier, ConsumingClassifier),
-        (MultiOutputRegressor, ConsumingRegressor),
-    ],
-    ids=["Classifier", "Regressor"],
+    "metaestimator",
+    METAESTIMATORS,
+    ids=[str(row["metaestimator"].__class__.__name__) for row in METAESTIMATORS],
 )
-def test_multioutput_metadata_routing(MultiOutput, Estimator):
-    # Check routing of metadata
-    metaest = MultiOutput(Estimator())
-    warn_msg = (
-        "You are passing metadata for which the request values are not explicitly"
-        " set: sample_weight, metadata."
-    )
-    with pytest.warns(FutureWarning, match=(warn_msg)):
-        metaest.fit(X, y_multi, sample_weight=sample_weight, metadata=metadata)
-        check_recorded_metadata(
-            metaest.estimators_[0],
-            "fit",
-            sample_weight=sample_weight,
-            metadata=metadata,
+def test_warning_for_indicated_methods(metaestimator):
+    # Check that the indicated methods give a warning
+    # TODO: After deprecation period, always error
+    cls = metaestimator["metaestimator"]
+    registry = _Registry()
+    estimator = metaestimator["estimator"](registry=registry)
+    estimator_name = metaestimator["estimator_name"]
+    instance = cls(**{estimator_name: estimator})
+    X = metaestimator["X"]
+    y = metaestimator["y"]
+    routing_methods = metaestimator["routing_methods"]
+
+    for method_name in routing_methods:
+        if not metaestimator.get(method_name, {}).get("warns_on"):
+            # this method is not expected to warn
+            continue
+
+        warn_msg = (
+            "You are passing metadata for which the request values are not explicitly "
+            "set: sample_weight, metadata. From version 1.4 this results in the "
+            "following error: [sample_weight, metadata] are passed but are not "
+            "explicitly set as requested or not for "
+            f"{estimator.__class__.__name__}.{method_name}"
         )
+        with pytest.warns(FutureWarning, match=re.escape(warn_msg)):
+            method = getattr(instance, method_name)
+            method(X, y, sample_weight=sample_weight, metadata=metadata)
```

Comment:
this assumes only `fit` and `partial_fit` route things around, but other methods could do the same, like `transform`, `score` etc.

Reply:
Do you mean the assumption is implicitly made because of how the method is called? I.e. we need a more generic way to call the method? If so, what needs to be generic: calling with `y`, calling with `sample_weight`?

Reply:
the way the method is called is fine, but if the method being called is not `fit` or `partial_fit`, it'll raise an exception that the estimator has not been `fit`.

Reply:
Hmm, I see. I have no really good solution for this that would work in all cases. Just something OTOH:

```python
# before calling method
if not "fit" in method_name:
    instance.fit(X, y)  # <= would probably still fail on some transformers
...
```

Reply:
yeah that would work I think.

Reply:
Done